lib/decode.js

```diff
@@ -0,0 +1,143 @@
+    const appTermMap = cborMap.value.get(0);
+    cborMap.value.delete(0);
+    const termCodecMap = await getTermCodecMap(
+      {contextUrls, appTermMap, documentLoader, decode: true});
+
+    // generate the decoding Map from the CBOR map
+    const decodingMap = _generateDecodingMap({
+      cborMap: cborMap.value, appContextMap, appTermMap, termCodecMap});
+
+    jsonldDocument = _decodeMap(decodingMap);
+  } else {
+    throw new CborldError('ERR_NOT_CBORLD',
+      'Input is not valid CBOR-LD; missing CBOR-LD header (0x0500 or 0x501).');
+  }
+
+  if(diagnose) {
+    diagnose('Diagnostic CBOR Object:');
+    diagnose(await util.inspect(cborMap, {depth: null, colors: true}));
+    diagnose('Diagnostic JSON-LD Object:');
+    diagnose(await util.inspect(jsonldDocument, {depth: null, colors: true}));
+  }
+
+  return jsonldDocument;
+}
+
+// generate a CBOR encoding map given a JSON-LD document and term codec map
+// FIXME: This largely duplicates code from encode.js/_generateEncodingMap
+function _generateDecodingMap(
+  {cborMap, appContextMap, termCodecMap}) {
+  const decodingMap = new Map();
+
+  for(const [key, value] of cborMap.entries()) {
+    // check for undefined term codec keys
+    if(termCodecMap && termCodecMap.get(key) === undefined) {
+      throw new CborldError(
+        'ERR_UNKNOWN_CBORLD_TERM',
+        `Unknown term '${key}' was detected in JSON-LD input.`);
+    }
+
+    // get the decoded term codec
+    const term = (termCodecMap) ?
+      termCodecMap.get(key).value : key;
+
+    // set the encoded value
+    let decodedValue = value;
+    if(value instanceof Map) {
+      decodedValue = _generateDecodingMap({
+        cborMap: value, appContextMap, termCodecMap});
+    } else if(termCodecMap) {
+      const Codec = termCodecMap.get(key).codec;
+      decodedValue = new Codec();
+      decodedValue.set({value, appContextMap, termCodecMap});
+    }
+
+    decodingMap.set(term, decodedValue);
+  }
+
+  return decodingMap;
+}
+
+// decodes a CBOR-LD decoding map
+function _decodeMap(decodingMap) {
+  const jsonldDocument = {};
+
+  for(const [key, value] of decodingMap.entries()) {
+    // set the encoded value
+    let decodedValue = value;
+    if(value instanceof Map) {
+      decodedValue = _decodeMap(value);
+    } else if(Array.isArray(value)) {
+      decodedValue = [];
+      value.forEach(element => {
+        let decodedSubvalue = element;
+        // FIXME: Repetitive code, doesn't work for arrays of arrays?
+        if(value instanceof Map) {
+          decodedSubvalue = _decodeMap(element);
+        } else if(element.decodeCBOR) {
+          decodedSubvalue = element.decodeCBOR();
+        } else {
+          throw new CborldError(
+            'ERR_UNKNOWN_CBORLD_ENCODING',
+            `Unknown encoding for '${key}' detected in CBOR-LD input: ` +
+            `${util.inspect(element)}`);
+        }
+        decodedValue.push(decodedSubvalue);
+      });
```

Comment:
So your FIXME suggests this doesn't work:

```suggestion
      decodedValue = value.map(_decodeMap);
```

If so, please ignore, it can be fixed later.

Reply:
I don't think this works because of the `else if(element.decodeCBOR) {` statement...